### PR TITLE
Refactor PrometheusProviders

### PR DESCRIPTION
**Pull request by Nokel81**
~~Don't merge until #2712 has been merged to master~~

- Remove implementation leak by passing in query name

- Remove renderer assuming static list by providing information from main

- Fix up cluster features (a bit) so that they properly track an individual cluster

- Catch errors from installing lens metrics, and propagate the errors from main to renderer

Signed-off-by: Sebastian Malton <sebastian@malton.name>

### CODE REVIEW

File: src/extensions/cluster-feature.ts [hunk 1]
```diff
@@ -39,29 +39,31 @@ export abstract class ClusterFeature {
     canUpgrade: false
   };
 
+  constructor(protected cluster: KubernetesCluster){}
+
   /**
    * to be implemented in the derived class, this method is typically called by Lens when a user has indicated that this feature is to be installed. The implementation
    * of this method should install kubernetes resources using the applyResources() method, or by directly accessing the kubernetes api (K8sApi)
    *
    * @param cluster the cluster that the feature is to be installed on
```
Comment (jim-docker) [inline src/extensions/cluster-feature.ts]: remove this (and similarly below)

File: src/main/router.ts [hunk 2]
```diff
@@ -5,7 +5,7 @@ import path from "path";
 import { readFile } from "fs-extra";
 import { Cluster } from "./cluster";
 import { apiPrefix, appName, publicPath, isDevelopment, webpackDevServerPort } from "../common/vars";
-import { helmRoute, kubeconfigRoute, metricsRoute, portForwardRoute, resourceApplierRoute, versionRoute } from "./routes";
+import { HelmApiRoute, KubeconfigRoute, MetricsRoute, PortForwardRoute, ResourceApplierApiRoute, VersionRoute } from "./routes";
```
Comment (jim-docker) [inline src/main/router.ts]: This PR is pretty big, can these router changes be done in another PR?

File: src/renderer/components/cluster-settings/components/cluster-prometheus-setting.tsx [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
-import { observable, computed, autorun } from "mobx";
+import { observable, autorun, computed } from "mobx";
+import { metricsApi } from "../../../api/endpoints/metrics.api";
+import { Spinner } from "../../spinner";
 
-const options: SelectOption<string>[] = [
-  { value: "", label: "Auto detect" },
-  ...prometheusProviders.map(pp => ({value: pp.id, label: pp.name}))
-];
+interface MetricProviderInfo {
```
Comment (jim-docker) [inline src/renderer/components/cluster-settings/components/cluster-prometheus-setting.tsx]: This is already defined in `metrics.api.ts`

File: extensions/metrics-cluster-feature/renderer.tsx [hunk 4]
```diff
@@ -12,42 +24,73 @@ export default class ClusterMetricsFeatureExtension extends LensRendererExtensio
...
         onClick: async () => {
-          metricsFeature.install(cluster);
+          try {
+            Component.Notifications.info(`Lens Metrics is being installed to ${cluster.metadata.name}`, { timeout: 7_500 });
 
-          Component.Notifications.info(`Lens Metrics is now installed to ${cluster.metadata.name}`, { timeout: 10_000 });
+            await metricsFeature.install();
+            await poll(() => metricsFeature.updateStatus(), ({ installed }) => installed, 5_000);
+
+            Component.Notifications.info(`Lens Metrics has been installed to ${cluster.metadata.name}`, { timeout: 7_500 });
```
Comment (jim-docker) [inline extensions/metrics-cluster-feature/renderer.tsx]: Minor UX thing: Is there a way to dismiss the first notification ("Lens Metrics is being installed...") when posting this one? It looks a little odd/cluttered to see both at the same time (when it takes less than 7.5 seconds to install)

File: src/main/router.ts (hunk 2 above)
Comment (Nokel81) [reply src/main/router.ts]: Good idea, will do

File: extensions/metrics-cluster-feature/renderer.tsx (hunk 4 above)
Comment (Nokel81) [reply extensions/metrics-cluster-feature/renderer.tsx]: Yes we can

**Comment by nevalla** [comment]
Just curious were there some issue with `LensApi` where `respondJson` and `respondText` are now exported as function?

**Comment by Nokel81** [comment]
> Just curious were there some issue with `LensApi` where `respondJson` and `respondText` are now exported as function?

Do you mean in the past? Because to me there are just helper functions are they work entirely on their inputs

**Review by jim-docker — DISMISSED** [review]

File: src/extensions/cluster-feature.ts [hunk 5]
```diff
@@ -39,96 +39,96 @@ export abstract class ClusterFeature {
     canUpgrade: false
   };
 
+  constructor(protected cluster: KubernetesCluster){}
```
Comment (jim-docker) [inline src/extensions/cluster-feature.ts]: Are we tracking breaking changes?

Comment (Nokel81) [reply src/extensions/cluster-feature.ts]: We should be, @jakolehm is going to be change this even more (I think eventually removing this class all together). We should have a doc somewhere for keeping track of all the breaking changes.

**Comment by nevalla** [comment]
Good job! I think after these changes it would be relatively easy to expose provider registry to extension api?

File: Makefile [hunk 6]
```diff
@@ -98,6 +98,7 @@ copy-extension-themes:
 src/extensions/npm/extensions/__mocks__:
 	cp -r __mocks__ src/extensions/npm/extensions/
 
+.PHONY: src/extensions/npm/extensions/dist
```
Comment (jakolehm) [inline Makefile]: Could you revert this change?

**Comment by Nokel81** [comment]
@nevalla Yes it should be, since anything that behaves like these black boxes should(tm) work

**Comment by Nokel81** [comment]
@jakolehm PTAL

**Review by jakolehm — COMMENTED** [review]
LGTM (expect Makefile change)

File: Makefile (hunk 6 above)
Comment (Nokel81) [reply Makefile]: Sure, though I always find that this correct, namely that the extensions should almost always be recreated.

Comment (Nokel81) [reply Makefile]: reverted

File: src/main/context-handler.ts [hunk 7]
```diff
@@ -72,18 +72,34 @@ export class ContextHandler {
...
+    }
+
+    return Array.from(registry.providers.values());
+  }
+
+  async getPrometheusService(): Promise<PrometheusService | undefined> {
+    const providers = this.listPotentialProviders();
+    const proxyConfig = await this.cluster.getProxyKubeconfig();
+    const apiClient = proxyConfig.makeApiClient(CoreV1Api);
+    const potentialServices = await Promise.allSettled(
```
Comment (jakolehm) [inline src/main/context-handler.ts:92]: Does this alter priority of potential services? (test that confirms would be nice)

**Review by jakolehm — DISMISSED** [review]

File: src/main/context-handler.ts (hunk 7 above)
Comment (Nokel81) [reply src/main/context-handler.ts:92]: It does not, it just means that if one of the `provider.getPrometheusService()` calls rejects, we can still proceed (namely it is skipped).

Comment (Nokel81) [reply src/main/context-handler.ts:92]: Test added.

**Review by nevalla — DISMISSED** [review]
Tested and seems to work

File: src/main/__test__/context-handler.test.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar");
+    });
+
+    it("should pick the first provider of 1 success, after 2 fails", async () => {
+      PrometheusProviderRegistry.getInstance()
+        .registerProvider(new TestProvider("id_foobar_1", AlwaysFail.Throw))
+        .registerProvider(new TestProvider("id_foobar_2", AlwaysFail.No))
```
Comment (jim-docker) [inline src/main/__test__/context-handler.test.ts]: Did you mean
```suggestion
        .registerProvider(new TestProvider("id_foobar_2", AlwaysFail.Return))
```

File: src/main/__test__/context-handler.test.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar_3");
+    });
+
+    it("should pick the first provider of 1 success, between 2 fails", async () => {
+      PrometheusProviderRegistry.getInstance()
+        .registerProvider(new TestProvider("id_foobar_1", AlwaysFail.Throw))
+        .registerProvider(new TestProvider("id_foobar_2"))
+        .registerProvider(new TestProvider("id_foobar_3", AlwaysFail.No));
```
Comment (jim-docker) [inline src/main/__test__/context-handler.test.ts]: same here

File: src/main/__test__/context-handler.test.ts (hunk 8 above)
Comment (Nokel81) [reply src/main/__test__/context-handler.test.ts]: Yes I did.

File: src/main/__test__/context-handler.test.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar_2");
```
Comment (jim-docker) [inline src/main/__test__/context-handler.test.ts]: this should've failed then?

Comment (Nokel81) [reply src/main/__test__/context-handler.test.ts]: No this is still correct because `AlwaysFail.No` is the default, and by design we pick the first one that succeeds.

Comment (jim-docker) [reply src/main/__test__/context-handler.test.ts]: Then the above test should've failed?

Comment (Nokel81) [reply src/main/__test__/context-handler.test.ts]: Maybe? will investigate.

**Review by nevalla — DISMISSED** [review]

**Review by jim-docker — DISMISSED** [review]
Did you figure out why that test (`should pick the first provider of 1 success, after 2 fails`) passed before you corrected it? Was it not run? Otherwise it's not a good test, right?

**Comment by Nokel81** [comment]
@jim-docker That test was correct in passing because it picked the first one that succeeded, the fact that the third one also succeeded shouldn't change that.

File: src/main/__test__/context-handler.test.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+    it("should pick the first provider of 1 success, after 2 fails", async () => {
+      PrometheusProviderRegistry.getInstance()
+        .registerProvider(new TestProvider("id_foobar_1", ServiceResult.Failure))
+        .registerProvider(new TestProvider("id_foobar_2", ServiceResult.Undefined))
+        .registerProvider(new TestProvider("id_foobar_3", ServiceResult.Success));
+
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar_3");
```
Comment (jim-docker) [inline src/main/__test__/context-handler.test.ts]: Before you corrected this test the first one that succeeded would've been "id_foobar_2", so this line should've failed, right?

Comment (Nokel81) [reply src/main/__test__/context-handler.test.ts]: Oh right, I thought you were talking about the third test. Will investigate.

Comment (Nokel81) [reply src/main/__test__/context-handler.test.ts]: I don't know why this test previously was passing. But I have added many more tests, and even one to validate that that specific case doesn't happen.

**Review by jim-docker — APPROVED** [review]

**Review by nevalla — APPROVED** [review]
